lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -3860,21 +3860,130 @@ SPIRVValue *LLVMToSPIRVBase::transIntrinsicInst(IntrinsicInst *II,
+      return BM->addSelectInst(Cmp, Add, Max, BB);
+    }
+    if (IID == Intrinsic::usub_sat) {
+      // usub.sat(a, b) -> (a > b) ? a - b : 0
+      SPIRVValue *Sub =
+          BM->addBinaryInst(OpISub, Ty, FirstArgVal, SecondArgVal, BB);
+      SPIRVValue *Cmp = BM->addCmpInst(OpUGreaterThan, SPVBoolTy, FirstArgVal,
+                                       SecondArgVal, BB);
+      SPIRVValue *Zero = transValue(Constant::getNullValue(II->getType()), BB);
+      return BM->addSelectInst(Cmp, Sub, Zero, BB);
+    }
+
+    uint64_t NumBits = LLVMTy->getScalarSizeInBits();
+    SPIRVValue *Zero = transValue(Constant::getNullValue(II->getType()), BB);
+    SPIRVValue *Max = transValue(
+        Constant::getIntegerValue(LLVMTy, APInt::getSignedMaxValue(NumBits)),
+        BB);
+    SPIRVValue *Min = transValue(
+        Constant::getIntegerValue(LLVMTy, APInt::getSignedMinValue(NumBits)),
+        BB);
+    SPIRVValue *IsPositive =
+        BM->addCmpInst(OpSGreaterThan, SPVBoolTy, SecondArgVal, Zero, BB);
+    SPIRVValue *IsNegative =
+        BM->addCmpInst(OpSLessThan, SPVBoolTy, SecondArgVal, Zero, BB);
+
+    if (IID == Intrinsic::sadd_sat) {
+      // sadd.sat(a, b) -> if (b > 0) && a > MAX - b => overflow -> MAX
+      //                -> else if (b < 0) && a < MIN - b => overflow -> MIN
+      //                -> else a + b
+      // There two ways to represent such sequence in IR:
+      // 1. Via 2 branch instructions plus 'phi' instruction;
+      // 2. Via set of select instructions.
+      // The later was chosed because of the following consideration:
+      // speculative branch prediction will most likely consider 'if' statements
+      // here as always false falling through to 'a + b', and reversing it will
+      // cause performance degradation.
+      SPIRVValue *Add =
+          BM->addBinaryInst(OpIAdd, Ty, FirstArgVal, SecondArgVal, BB);
+
+      // check if (b > 0) && a > MAX - b condition
+      SPIRVValue *MaxSubB =
+          BM->addBinaryInst(OpISub, Ty, Max, SecondArgVal, BB);
+      SPIRVValue *CanPosOverflow =
+          BM->addCmpInst(OpSGreaterThan, SPVBoolTy, FirstArgVal, MaxSubB, BB);
+      SPIRVValue *PosOverflow = BM->addInstTemplate(
+          OpLogicalAnd, {CanPosOverflow->getId(), IsPositive->getId()}, BB,
+          SPVBoolTy);
+
+      // check if (b < 0) && MIN - b > a condition
+      SPIRVValue *MinSubB =
+          BM->addBinaryInst(OpISub, Ty, Min, SecondArgVal, BB);
+      SPIRVValue *CanNegOverflow =
+          BM->addCmpInst(OpSGreaterThan, SPVBoolTy, MinSubB, FirstArgVal, BB);
+      SPIRVValue *NegOverflow = BM->addInstTemplate(
+          OpLogicalAnd, {CanNegOverflow->getId(), IsNegative->getId()}, BB,
+          SPVBoolTy);
+
+      // do selects
+      SPIRVValue *FirstSelect = BM->addSelectInst(PosOverflow, Max, Add, BB);
+      return BM->addSelectInst(NegOverflow, Min, FirstSelect, BB);
+    }
+    // ssub.sat(a, b) -> if (b > 0) && a < MIN + b => overflow -> MIN
```

Comment:
This can be:

res = a-b
if (b<0 && a>res) res = MAX
if (b>0 && a<res) res = MIN

This makes it unnecessary to calculate "MAX + b" and "MIN + b"